### ISSUE TITLE
Avoid producing a new CNAB during relocation

### ISSUE THREAD

**Issue by glyn** [deislabs/cnab-spec#195]
The current CNAB spec requires image relocation to capture a mapping from original to new image references in the CNAB so that the invocation image can take this into account. In other words, relocation takes a CNAB as input and produces a modified CNAB as output. This won't work well with signing.

It would be better for relocation to avoid mutating the CNAB and, instead, to output a separate file which could then be signed (and the relocation step included in the software supply chain).

To support this, a new JSON schema will be defined to capture the _relocation mapping_ from original to new image references, both for invocation images and for other images. An example of such a relocation mapping is:
```
{
  "images": {
    "my-microservice": {
      "description": "my microservice",
      "image": "my.registry/microservice@sha256:cca460afa270d4c527981ef9ca4989346c56cf9b20217dcea37df1ece8120687",
      "originalImage": "gabrtv/microservice@sha256:cca460afa270d4c527981ef9ca4989346c56cf9b20217dcea37df1ece8120687",
      "digest": "sha256:cca460afa270d4c527981ef9ca4989346c56cf9b20217dcea37df1ece8120687"
    }
  },
  "invocationImages": [
    {
      "digest": "sha256:86959ecb500308ae523922eab84f2f94082f20b4e7bda84ce9be219f3f1b4e65",
      "image": "my.registry/helloworld:0.1.0",
      "originalImage": "technosophos/helloworld:0.1.0",
      "imageType": "docker"
    }
  ]
}
```

The relocation mapping may be passed to the runtime along with the CNAB when executing actions against the CNAB. (If relocation has not occurred, then the CNAB contains all the information that would be in the relocation mapping and a relocation mapping need not be passed to the runtime.)

The `invocationImages` section of the relocation mapping tells the runtime the location of relocated invocation images.

The `images` section of the relocation mapping tells the invocation image the location of relocated images. The bundle runtime chapter of the spec will be updated to say that the `/cnab/app/image-map.json` file mounted in the invocation image will contain the `images` section of the relocation mapping. In the above example, the mounted file would contain:
```
{
  "my-microservice": {
      "description": "my microservice",
      "image": "my.registry/microservice@sha256:cca460afa270d4c527981ef9ca4989346c56cf9b20217dcea37df1ece8120687",
      "originalImage": "gabrtv/microservice@sha256:cca460afa270d4c527981ef9ca4989346c56cf9b20217dcea37df1ece8120687",
      "digest": "sha256:cca460afa270d4c527981ef9ca4989346c56cf9b20217dcea37df1ece8120687"
  }
}
```

The `originalImage` fields in bundle.json are no longer needed and will be deleted.

**Comment by glyn** [comment]
If https://github.com/deislabs/cnab-spec/issues/144 is added to the spec, then the relocation mapping proposed above can be greatly simplified since it need only provide the actual mapping from original image references to relocated image references, probably without needing to differentiate between invocation and other images. The above relocation mapping could then be simplifed to:
```
{
 "gabrtv/microservice@sha256:cca460afa270d4c527981ef9ca4989346c56cf9b20217dcea37df1ece8120687": "my.registry/microservice@sha256:cca460afa270d4c527981ef9ca4989346c56cf9b20217dcea37df1ece8120687",
 "technosophos/helloworld:0.1.0": "my.registry/helloworld:0.1.0"
}
```

**Comment by glyn** [comment]
@radu-matei After Thursday 20th June, I will on vacation until after the v1.0 spec deadline, so if you'd like to discuss this issue with me, please grab me before then.

I think the spec changes need to include the following:
* Since https://github.com/deislabs/cnab-spec/issues/144 looks set to be addressed, relocating a CNAB should no longer produce a new CNAB but should produce a _relocation mapping file_ in the simplified format above.
* The relocation mapping file should then, optionally, be mounted in the runtime filesystem of the invocation image. If we need to say when, it should be mounted for install, upgrade, uninstall (not certain about this, but it's good for symmetry if nothing else), and custom actions.
* The `originalImage` field should be deleted from the spec, including examples, and the JSON schema, both for invocation images and other images.
* The image map should no longer be mounted at `/cnab/app/image-map.json` and the spec should be reworked correspondingly.
* `805-airgap.md` should be reworked to describe the above new approach to relocation.

**Comment by glyn** [comment]
Another option to consider is to regard the production and consumption of the relocation mapping file to be implementation specific and simply define the format of the file and where it should optionally be mounted when a bundle has been relocated. However, I don't see much advantage in that and I think it would make the spec unnecessarily obscure.

**Comment by simonferquel** [comment]
I really like this approach, thanks!

**Comment by glyn** [comment]
> * The relocation mapping file should include in its keys all the images defined by the CNAB.

Or, probably better, the relocation mapping MUST include in its keys all the images defined by the CNAB and MAY include mappings for [synonyms](https://github.com/pivotal/image-relocation/blob/1445a3d1e9b659ed2700a0ff537848a525505dbe/pkg/image/name_test.go#L201) of those keys.

With this, invocation images don't need to worry about synonyms,  since the keys of the relocation mapping will cover all the image references declared in the bundle.

**Comment by ryanmoran** [comment]
I got some feedback today from a team building a bundle and wanting to support the relocation use-case. They were wondering if runtimes could make a guarantee that there would always be a relocation mapping file in the invocation container. From there perspective, they didn't want to have to write code to first check if the file was there, and then to do the mapping.

@glyn and I talked about this privately yesterday. In that conversation, we identified that a runtime could inject an "identity" mapping in cases where relocation did not take place. For example:
- Relocation: `{ "a": "b" }`
- No relocation: `{ "a": "a" }`

I don't have a strong preference for either design choice. I just wanted to include some feedback I heard from bundle authors here.
